### DISCUSSION TITLE
Saleor-Platform developing

### DISCUSSION THREAD

**Discussion by cdw-diataal** [saleor/saleor#10044] [Q&A] [answered]
Hi, First of a thank you for you product. I'm trying to set up a webshop and want to use Saleor as the framework. I've cloned the saleor-platform locally and tried making some changes on the storefron after I run the "docker-compose up" command. I noticed that if i want to see my changes that i have to reload my container. Is there any way to activate the hot reloading for the storefront and dashboard when developing using the saleor-platform? Thank you

**Comment by karolkielecki** [comment]
Hi @cdw-diataal, I think it should work in the latest version. But we need to check it out -- we'll take a look at it soon as we can.

**Comment by timuric** [comment, marked as answer]
@cdw-diataal I would recommend developing Storefront and Dashboard outside of docker, it is pretty simple to setup, you just need to set env variable of the API (backend) endpoint.

**Reply by cdw-diataal** [reply]
Hi @timuric I tried starting the dashboard by following the instructions on the github page (npm i and npm start) but I get the following error:

```
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   'C:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'start'
1 verbose cli ]
2 info using npm@6.14.13
3 info using node@v14.17.0
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info lifecycle saleor-dashboard@3.4.0-dev~prestart: saleor-dashboard@3.4.0-dev
6 info lifecycle saleor-dashboard@3.4.0-dev~start: saleor-dashboard@3.4.0-dev
7 verbose lifecycle saleor-dashboard@3.4.0-dev~start: unsafe-perm in lifecycle true
8 verbose lifecycle saleor-dashboard@3.4.0-dev~start: PATH: C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin;D:\supplij_dev\dashboard\node_modules\.bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.6\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.6\libnvvp;C:\Program Files\Common Files\Oracle\Java\javapath;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\NVIDIA Corporation\NVIDIA NvDLISR;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\nodejs\;C:\Program Files\dotnet\;C:\Program Files\NVIDIA Corporation\Nsight Compute 2022.1.0\;C:\Program Files\Git\cmd;C:\Program Files\Docker\Docker\resources\bin;C:\ProgramData\DockerDesktop\version-bin;C:\Users\Gebruiker\AppData\Local\Programs\Python\Python39\Scripts\;C:\Users\Gebruiker\AppData\Local\Programs\Python\Python39\;C:\Users\Gebruiker\AppData\Local\Microsoft\WindowsApps;C:\Program Files\JetBrains\PyCharm 2020.3.1\bin;;C:\Program Files\JetBrains\IntelliJ IDEA 2020.3.1\bin;;C:\Users\Gebruiker\AppData\Roaming\JetBrains\PyCharm2020.3\node\node-v14.15.0-win-x64\node_modules\npm;C:\Users\Gebruiker\AppData\Local\Programs\Microsoft VS Code\bin;C:\sdk\flutter\bin;C:\Users\Gebruiker\AppData\Roaming\npm;C:\flutter_sdk\flutter;
9 verbose lifecycle saleor-dashboard@3.4.0-dev~start: CWD: D:\dashboard
10 silly lifecycle saleor-dashboard@3.4.0-dev~start: Args: [ '/d /s /c', 'npm run build-types && webpack-dev-server -d' ]
11 silly lifecycle saleor-dashboard@3.4.0-dev~start: Returned: code: 1  signal: null
12 info lifecycle saleor-dashboard@3.4.0-dev~start: Failed to exec start script
13 verbose stack Error: saleor-dashboard@3.4.0-dev start: `npm run build-types && webpack-dev-server -d`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:376:20)
13 verbose stack     at ChildProcess.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\lib\spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:376:20)
13 verbose stack     at maybeClose (internal/child_process.js:1055:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)
14 verbose pkgid saleor-dashboard@3.4.0-dev
15 verbose cwd D:\dashboard
16 verbose Windows_NT 10.0.22000
17 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "start"
18 verbose node v14.17.0
19 verbose npm  v6.14.13
20 error code ELIFECYCLE
21 error errno 1
22 error saleor-dashboard@3.4.0-dev start: `npm run build-types && webpack-dev-server -d`
22 error Exit status 1
23 error Failed at the saleor-dashboard@3.4.0-dev start script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

Do you have any idea what I am doing wrong?

**Reply by cdw-diataal** [reply]
I forgot to mention i also created a .env file containing the following:

```
API_URI=http://localhost:8000/graphql/
APP_MOUNT_URI=/dashboard/
STATIC_URL=/static/
```

**Reply by timuric** [reply]
@cdw-diataal which node and npm version are you using?

Could you please tell why do you want to fork dashboard instead of extending it?

**Reply by cdw-diataal** [reply]
@timuric I'm using the following:

2 info using npm@6.14.13
3 info using node@v14.17.0

the answer on why I forked it. Thats how my boss wants it but we do have a upstream to the original repository to get all the updates if needed.

**Reply by timuric** [reply]
@cdw-diataal could you please try node 18? You could use `nvm` to easily switch versions

**Reply by cdw-diataal** [reply]
@timuric I'll try that as soon as I can. I will let you know if that worked. Thank you

**Reply by tiendq** [reply]
@timuric 

I could clone and build storefront and dashboard successfully for local development, it's pretty simple since they're NodeJS/React SPA.

But how is about core API? Should I clone it too and setup everything on my macOS for local development or using `saleor-platform`? In case using `saleor-platform` how could I work with my fork e.g. to fix a bug? I don't see anything like this is mentioned clearly in Saleor docs (Contribution section).

Thanks,

**Reply by tiendq** [reply]
@cdw-diataal 

AFAIK dashboard (3.5.0) has restrictions for Node v14.x only, and when I removed these restrictions I saw many build errors on Node v16.x.

**Reply by peelar** [reply]
Hi @tiendq 👋 

> @timuric
> But how is about core API? Should I clone it too and setup everything on my macOS for local development or using `saleor-platform`? 

Quoting the [saleor-platform](https://github.com/saleor/saleor-platform) documentation:
> Keep in mind this repository is for local development only and is not meant to be deployed on any production environment! 

There is some complexity in this topic that we streamline with Saleor Cloud (where we take care of the deployment, among many others), but if you are planning to host Saleor on your own, you can find [an example of Heroku deployment](https://docs.saleor.io/docs/3.x/developer/running-saleor/heroku) in our documentation. I will make sure it's up to date today. We are also planning on documenting some other deployment options in the near future. Stay tuned 👂

**Reply by tiendq** [reply]
> In case using saleor-platform how could I work with my fork e.g. to fix a bug? I don't see anything like this is mentioned clearly in Saleor docs (Contribution section).

Thanks @peelar for your reply but it seems you missed this part, I'm curious to know how other contributors work with this repo (`saleor` core), normally following GitHub guidelines we should forking the repo first, make changes, and submit a PR.

**Reply by peelar** [reply]
@tiendq

First of all, deep thanks for your interest in contributing to the Saleor community 🙇. The process of chiming into this repo is exactly as you described: you fork it, make your changes, and submit a PR.

**Comment by PlusA2M** [comment]
Hi @cdw-diataal ! We're currently actively developing Saleor plugin for our own use, the `api` itself is capable to hot reload, it's has a `autoreload` mechanism built right in. I also agree with @timuric that you may separate `api`, `storefront` and `dashboard`, only `api` would require developing inside a docker, due to the fact that we don't need the storefront, we never spinning up the storefront also, spinning up either the `dashboard` or `storefront` would significantly cost time than `api`.

**Reply by timuric** [reply]
I would highly recommend you to explore authoring apps instead of plugins

**Reply by PlusA2M** [reply]
Hi @timuric ! Thanks for your suggestion, I've heard that Saleor may drop support for plug-in, hopefully it will not be deprecated soon (since we've finished our beta version of the development🤣). We think the nature and feature provided by plug-in mechanism would better fit our needs at the moment, we've also seen the new iFrame feature provided through apps, we're so interested also. Our use case here fyr: [#7718](https://github.com/saleor/saleor/discussions/7718). Any idea or thoughts are also welcome!

**Reply by timuric** [reply]
@PlusA2M  Let me know if this solves your problem https://github.com/saleor/saleor/discussions/7718#discussioncomment-3028712

**Comment by cdw-diataal** [comment]
Aaah I see, then I'll comment out the dashboard and frontend and run those separately, Thank you for your answers all

**Comment by RGuven** [comment]
@timuric Hi, 
is there any improvement for hot reloading? 
I can send PR if no improvement has been made.

**Comment by varada13** [comment]
i setup saleor platform according to docs, and after i run 
docker-compose up
core, dashboard are running in 8000 and 9000 ports but,
client storefront is not running , what shoud i do?
